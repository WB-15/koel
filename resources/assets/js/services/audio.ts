export const audio = {
  context: null as AudioContext | null,
  source: null as MediaElementAudioSourceNode | null,
  element: null as HTMLMediaElement | null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext (): AudioContext {
    return this.context!
  },

  getSource (): MediaElementAudioSourceNode {
    return this.source!
  },

  getElement (): HTMLMediaElement {
    return this.element!
  }
}
