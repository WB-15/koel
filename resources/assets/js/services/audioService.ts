export const audioService = {
  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,

  init (element: HTMLMediaElement) {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context
  },

  getSource () {
    return this.source
  },

  getElement () {
    return this.element
  }
}
