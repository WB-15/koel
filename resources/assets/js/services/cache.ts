const CACHE_EXPIRATION_TIME = 1000 * 60 * 60 * 24 // 1 day

export const Cache = {
  storage: new Map<any, {
    time: number,
    value: any
  }>(),

  normalizeKey: (key: any) => typeof key === 'object' ? JSON.stringify(key) : key,

  has (key: any) {
    return this.hit(this.normalizeKey(key))
  },

  set (key: any, value: any) {
    this.storage.set(this.normalizeKey(key), {
      value,
      time: Date.now()
    })
  },

  hit (key: any) {
    return !this.miss(this.normalizeKey(key))
  },

  miss (key: any) {
    key = this.normalizeKey(key)

    if (!this.storage.has(key)) return true
    const { time } = this.storage.get(key)!

    if (time < Date.now() - CACHE_EXPIRATION_TIME) {
      this.storage.delete(key)
      return true
    }

    return false
  },

  invalidate (key: any) {
    this.storage.delete(this.normalizeKey(key))
  },

  resolve<T> (key: any, resolver: Closure) {
    key = this.normalizeKey(key)

    this.hit(key) || this.set(key, resolver())
    return this.storage.get(key)!.value
  }
}
