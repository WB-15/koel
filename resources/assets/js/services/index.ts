export * from './httpService'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
