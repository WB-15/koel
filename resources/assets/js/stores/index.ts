export * from './albumStore'
export * from './artistStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './playlistStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './settingStore'
export * from './commonStore'
export * from './songStore'
export * from './userStore'
export * from './searchStore'
export * from './themeStore'
