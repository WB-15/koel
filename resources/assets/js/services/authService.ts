import { localStorageService } from '@/services'

const STORAGE_KEY = 'api-token'

export const authService = {
  getToken: () => localStorageService.get<string | null>(STORAGE_KEY),

  hasToken () {
    return Boolean(this.getToken())
  },

  setToken: (token: string) => localStorageService.set(STORAGE_KEY, token),
  destroy: () => localStorageService.remove(STORAGE_KEY)
}
