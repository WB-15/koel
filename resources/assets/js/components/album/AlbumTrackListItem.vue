<template>
  <li :class="{ available: song }" :title="tooltip" role="button" tabindex="0" @click="play">
    <span class="title">{{ track.title }}</span>
    <a
      v-if="useiTunes && !song"
      :href="iTunesUrl"
      class="view-on-itunes"
      target="_blank"
      title="View on iTunes"
    >
      iTunes
    </a>
    <span class="length">{{ track.fmtLength }}</span>
  </li>
</template>

<script lang="ts" setup>
import { computed, toRef, toRefs } from 'vue'
import { queueStore, commonStore, songStore } from '@/stores'
import { authService, playbackService } from '@/services'

const props = defineProps<{ album: Album, track: AlbumTrack }>()
const { album, track } = toRefs(props)

const useiTunes = toRef(commonStore.state, 'useiTunes')

const song = computed(() => songStore.guess(track.value.title, album.value))
const tooltip = computed(() => song.value ? 'Click to play' : '')

const iTunesUrl = computed(() => {
  return `${window.BASE_URL}itunes/song/${album.value.id}?q=${encodeURIComponent(track.value.title)}&api_token=${authService.getToken()}`
})

const play = () => {
  if (song.value) {
    queueStore.queueIfNotQueued(song.value)
    playbackService.play(song.value)
  }
}
</script>

<style lang="scss" scoped>
[role=button] {
  &:focus {
    span.title {
      color: var(--color-highlight);
    }
  }

  a.view-on-itunes {
    display: inline-block;
    border-radius: 3px;
    font-size: .8rem;
    padding: 0 5px;
    color: var(--color-text-primary);
    background: rgba(255, 255, 255, .1);
    height: 20px;
    line-height: 20px;
    margin-left: 4px;

    &:hover, &:focus {
      background: linear-gradient(27deg, #fe5c52 0%, #c74bd5 50%, #2daaff 100%);
      color: var(--color-text-primary);
    }

    &:active {
      box-shadow: inset 0px 5px 5px -5px #000;
    }
  }
}
</style>
