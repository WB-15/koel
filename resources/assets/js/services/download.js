import $ from 'jquery';
import { map } from 'lodash';

import { playlistStore, favoriteStore } from '../stores';
import { ls } from '.';

export const download = {
    /**
     * Download individual song(s).
     *
     * @param {Array.<Object>|Object} songs
     */
    fromSongs(songs) {
        songs = [].concat(songs);
        const ids = map(songs, 'id');
        const params = $.param({ songs: ids });

        return this.trigger(`songs?${params}`);
    },

    /**
     * Download all songs in an album.
     *
     * @param {Object} album
     */
    fromAlbum(album) {
        return this.trigger(`album/${album.id}`);
    },

    /**
     * Download all songs performed by an artist.
     *
     * @param {Object} artist
     */
    fromArtist(artist) {
        // It's safe to assume an artist always has songs.
        // After all, what's an artist without her songs?
        // (See what I did there? Yes, I'm advocating for women's rights).
        return this.trigger(`artist/${artist.id}`);
    },

    /**
     * Download all songs in a playlist.
     *
     * @param {Object} playlist
     */
    fromPlaylist(playlist) {
        if (!playlistStore.getSongs(playlist).length) {
            console.warn('Empty playlist.');
            return;
        }

        return this.trigger(`playlist/${playlist.id}`);
    },

    /**
     * Download all favorite songs.
     */
    fromFavorites() {
        if (!favoriteStore.all.length) {
            console.warn("You don't like any song? Come on, don't be that grumpy.");
            return;
        }

        return this.trigger('favorites');
    },

    /**
     * Build a download link using a segment and trigger it.
     *
     * @param  {string} uri The uri segment, corresponding to the song(s),
     *                      artist, playlist, or album.
     */
    trigger(uri) {
        const sep = uri.indexOf('?') === -1 ? '?' : '&';
        const frameId = `downloader${Date.now()}`;
        $(`<iframe id="${frameId}" style="display:none"></iframe`).appendTo('body');
        document.getElementById(frameId).src = `/api/download/${uri}${sep}jwt-token=${ls.get('jwt-token')}`;
    },
}
