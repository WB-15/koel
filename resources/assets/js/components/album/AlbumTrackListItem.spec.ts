import { cleanup, fireEvent } from '@testing-library/vue'
import { beforeEach, expect, it } from 'vitest'
import factory from '@/__tests__/factory'
import { mockHelper, render } from '@/__tests__/__helpers__'
import { commonStore, queueStore, songStore } from '@/stores'
import { playbackService } from '@/services'
import AlbumTrackListItem from './AlbumTrackListItem.vue'

let song: Song

const track = {
  title: 'Fahrstuhl to Heaven',
  fmtLength: '00:42'
}

const album = factory<Album>('album', { id: 42 })

beforeEach(() => {
  cleanup()
  mockHelper.restoreAllMocks()

  song = factory<Song>('song')
  commonStore.state.useiTunes = true
})

it('renders', () => {
  const { html } = render(AlbumTrackListItem, {
    props: {
      album,
      track
    }
  })

  expect(html()).toMatchSnapshot()
})

it('plays', async () => {
  const guessMock = mockHelper.mock(songStore, 'guess', song)
  const queueMock = mockHelper.mock(queueStore, 'queueIfNotQueued')
  const playMock = mockHelper.mock(playbackService, 'play')

  const { getByTitle } = render(AlbumTrackListItem, {
    props: {
      album,
      track
    }
  })

  await fireEvent.click(getByTitle('Click to play'))

  expect(guessMock).toHaveBeenCalledWith('Fahrstuhl to Heaven', album)
  expect(queueMock).toHaveBeenNthCalledWith(1, song)
  expect(playMock).toHaveBeenNthCalledWith(1, song)
})
