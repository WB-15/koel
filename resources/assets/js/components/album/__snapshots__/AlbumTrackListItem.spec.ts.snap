// Vitest Snapshot v1

exports[`renders 1`] = `"<li class=\\"\\" title=\\"\\" tabindex=\\"0\\" data-v-d8ad5538=\\"\\"><span class=\\"title\\" data-v-d8ad5538=\\"\\">Fahrstuhl to Heaven</span><a href=\\"undefineditunes/song/42?q=Fahrstuhl%20to%20Heaven&amp;api_token=null\\" class=\\"view-on-itunes\\" target=\\"_blank\\" title=\\"View on iTunes\\" data-v-d8ad5538=\\"\\"> iTunes </a><span class=\\"length\\" data-v-d8ad5538=\\"\\">00:42</span></li>"`;
